applications/MappingApplication/custom_utilities/projection_utilities.cpp

```diff
@@ -0,0 +1,273 @@
+        double edge_distance;
+
+        for (const auto& r_edge : const_cast<GeometryType&>(rGeometry).Edges()) { // TODO update this after #4767
+
+            const PairingIndex edge_index = ProjectOnLine(r_edge, rPointToProject, LocalCoordTol, edge_sf_values, edge_eq_ids, edge_distance);
+
+            // check if the current edge gives a better result
+            if (edge_index > pairing_index || (edge_index == pairing_index && edge_distance < rProjectionDistance)) {
+                pairing_index = edge_index;
+                rShapeFunctionValues = edge_sf_values;
+                rProjectionDistance = edge_distance;
+                rEquationIds = edge_eq_ids;
+            }
+        }
+    }
+
+    return pairing_index;
+}
+
+PairingIndex ProjectIntoVolume(const GeometryType& rGeometry,
+                               const Point& rPointToProject,
+                               const double LocalCoordTol,
+                               Vector& rShapeFunctionValues,
+                               std::vector<int>& rEquationIds,
+                               double& rProjectionDistance,
+                               const bool ComputeApproximation)
+{
+    array_1d<double, 3> local_coords;
+    PairingIndex pairing_index;
+    bool is_inside = rGeometry.IsInside(rPointToProject, local_coords);
+
+    if (is_inside) {
+        pairing_index = PairingIndex::Volume_Inside;
+        rGeometry.ShapeFunctionsValues(rShapeFunctionValues, local_coords);
+        FillEquationIdVector(rGeometry, rEquationIds);
+
+        rProjectionDistance = MapperUtilities::ComputeDistance(rPointToProject, rGeometry.Center());
+        rProjectionDistance /= rGeometry.Volume(); // Normalize Distance by Volume
+
+    } else if (!ComputeApproximation) {
+        return PairingIndex::Unspecified;
+
+    } else if (!is_inside && rGeometry.IsInside(rPointToProject, local_coords, LocalCoordTol)) {
+        pairing_index = PairingIndex::Volume_Outside;
+        rGeometry.ShapeFunctionsValues(rShapeFunctionValues, local_coords);
+        FillEquationIdVector(rGeometry, rEquationIds);
+
+        rProjectionDistance = MapperUtilities::ComputeDistance(rPointToProject, rGeometry.Center());
+        rProjectionDistance /= rGeometry.Volume(); // Normalize Distance by Volume
+
+    } else { // inter-/extrapolation failed, trying to project on "subgeometries"
+        pairing_index = PairingIndex::Unspecified;
+        std::vector<int> face_eq_ids;
+        Vector face_sf_values;
+        double face_distance;
+
+        for (const auto& r_face : const_cast<GeometryType&>(rGeometry).Faces()) { // TODO update this after #4767
+
+            const PairingIndex face_index = ProjectOnSurface(r_face, rPointToProject, LocalCoordTol, face_sf_values, face_eq_ids, face_distance);
+
+            // check if the current edge gives a better result
+            if (face_index > pairing_index || (face_index == pairing_index && face_distance < rProjectionDistance)) {
```

Comment:
The first condition is rather obscure. I would suggest some named method to do the check, although I don't have great ideas. `IsMoreSpecificThan`? `IsBetterMatch`?